File: backend/internal/domain/invitation/invitation.go
package invitation

import (
	"context"
	"time"
)

// Invitation represents a pending organization member invitation
type Invitation struct {
	ID             int64      `gorm:"primaryKey" json:"id"`
	OrganizationID int64      `gorm:"not null;index" json:"organization_id"`
	Email          string     `gorm:"size:255;not null;index" json:"email"`
	Role           string     `gorm:"size:20;not null;default:member" json:"role"`
	Token          string     `gorm:"size:255;not null;uniqueIndex" json:"-"`
	InvitedBy      int64      `gorm:"not null" json:"invited_by"`
	ExpiresAt      time.Time  `gorm:"not null" json:"expires_at"`
	AcceptedAt     *time.Time `json:"accepted_at,omitempty"`
	CreatedAt      time.Time  `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt      time.Time  `gorm:"autoUpdateTime" json:"updated_at"`
}

// TableName specifies the table name for GORM
func (Invitation) TableName() string {
	return "invitations"
}

// IsExpired checks if the invitation has expired
func (i *Invitation) IsExpired() bool {
	return time.Now().After(i.ExpiresAt)
}

// IsAccepted checks if the invitation has been accepted
func (i *Invitation) IsAccepted() bool {
	return i.AcceptedAt != nil
}

// IsPending checks if the invitation is still pending
func (i *Invitation) IsPending() bool {
	return !i.IsAccepted() && !i.IsExpired()
}

// AcceptInvitationParams holds parameters for the atomic accept operation
type AcceptInvitationParams struct {
	Invitation *Invitation
	UserID     int64
	Role       string
}

// AcceptInvitationResult holds the result of accepting an invitation
type AcceptInvitationResult struct {
	OrganizationID int64
	MemberID       int64
}

// OrgInfo holds basic organization information
type OrgInfo struct {
	Name string
	Slug string
}

// Repository defines the interface for invitation data access
type Repository interface {
	Create(ctx context.Context, invitation *Invitation) error
	GetByToken(ctx context.Context, token string) (*Invitation, error)
	GetByID(ctx context.Context, id int64) (*Invitation, error)
	GetByOrgAndEmail(ctx context.Context, orgID int64, email string) (*Invitation, error)
	ListByOrganization(ctx context.Context, orgID int64) ([]*Invitation, error)
	ListPendingByEmail(ctx context.Context, email string) ([]*Invitation, error)
	Update(ctx context.Context, invitation *Invitation) error
	Delete(ctx context.Context, id int64) error
	DeleteExpired(ctx context.Context) error

	// Cross-entity queries (avoid Service needing direct DB access)
	CheckMemberExists(ctx context.Context, orgID int64, userID int64) (bool, error)
	CheckMemberExistsByEmail(ctx context.Context, orgID int64, email string) (bool, error)
	GetOrganization(ctx context.Context, orgID int64) (*OrgInfo, error)
	GetUserDisplayName(ctx context.Context, userID int64) (string, error)

	// AcceptInvitationAtomic atomically adds a member and marks the invitation as accepted
	AcceptInvitationAtomic(ctx context.Context, params *AcceptInvitationParams) (*AcceptInvitationResult, error)
}

File: web/src/lib/api/invitation.ts
import { request, orgPath } from "./base";

// Invitation types
export interface Invitation {
  id: number;
  organization_id: number;
  email: string;
  role: "admin" | "member";
  expires_at: string;
  accepted_at?: string;
  created_at: string;
}

export interface InvitationInfo {
  id: number;
  email: string;
  role: string;
  organization_id: number;
  organization_name: string;
  organization_slug: string;
  inviter_name: string;
  expires_at: string;
  is_expired: boolean;
}

export interface PendingInvitation {
  id: number;
  organization_id: number;
  organization_name: string;
  organization_slug: string;
  role: string;
  expires_at: string;
  token: string;
}

// Invitation API
export const invitationApi = {
  // Organization-scoped routes (require org context via X-Organization-Slug header)
  list: () =>
    request<{ invitations: Invitation[] }>(orgPath("/invitations")),

  create: (email: string, role: "admin" | "member") =>
    request<{ message: string; invitation: Invitation }>(orgPath("/invitations"), {
      method: "POST",
      body: { email, role },
    }),

  revoke: (id: number) =>
    request<{ message: string }>(`${orgPath("/invitations")}/${id}`, {
      method: "DELETE",
    }),

  resend: (id: number) =>
    request<{ message: string }>(`${orgPath("/invitations")}/${id}/resend`, {
      method: "POST",
    }),

  // Public/auth routes
  getByToken: (token: string) =>
    request<{ invitation: InvitationInfo }>(`/api/v1/invitations/${token}`),

  accept: (token: string) =>
    request<{ message: string; organization: { id: number; name: string; slug: string } }>(
      `/api/v1/invitations/${token}/accept`,
      { method: "POST" }
    ),

  listPending: () =>
    request<{ invitations: PendingInvitation[] }>("/api/v1/invitations/pending"),
};
